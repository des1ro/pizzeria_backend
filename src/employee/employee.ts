export class Employee {
  constructor(readonly name: string, readonly role: EmployeeRole) {}
}
