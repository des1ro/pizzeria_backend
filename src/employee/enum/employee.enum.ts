export enum EmployeeRole {
  Cheff = "Cheff",
  Waiter = "Waiter",
}
