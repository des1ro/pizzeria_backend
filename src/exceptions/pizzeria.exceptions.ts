type ErrorName =
  | "PERMISION_DENIED"
  | "TABLE_ERROR"
  | "ORDER_ERROR"
  | "EMPLOYEE_ERROR";
export class PizzeriaError extends Error {
  name: string;
  message: string;
  constructor({ name, message }: { name: ErrorName; message: string }) {
    super(name);
    this.name = name;
    this.message = message;
  }
}
