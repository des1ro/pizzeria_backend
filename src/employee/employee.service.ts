import { PizzeriaError } from "../exceptions/pizzeria.exceptions";
import { Employee } from "./employee";

export class EmployeeService {
  private readonly takenCheffs = new Set<Employee>();
  constructor(
    private readonly availableCheffs = new Set<Employee>(),
    private readonly availableWaiters = new Set<Employee>()
  ) {}
  getCheff() {
    const cheff = Array.from(this.availableCheffs).reverse().pop();
    if (cheff) {
      this.takenCheffs.add(cheff);
      return;
    }
    throw new PizzeriaError({
      name: "EMPLOYEE_ERROR",
      message: "There isn't free cheffs",
    });
  }
  releaseCheff(cheff: Employee) {
    if (this.takenCheffs.delete(cheff)) {
      this.availableCheffs.add(cheff);
      return;
    }
    throw new PizzeriaError({
      name: "EMPLOYEE_ERROR",
      message: "This cheff wasn't taken",
    });
  }
}
