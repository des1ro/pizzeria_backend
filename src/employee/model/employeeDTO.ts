import { EmployeeRole } from "../enum/employee.enum";

export class EmployeeDTO {
  constructor(readonly name: string, readonly role: EmployeeRole) {}
}
