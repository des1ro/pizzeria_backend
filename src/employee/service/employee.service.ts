import { EmployeeRole } from "../enum/employee.enum";
import { EmployeeDTO } from "../model/employeeDTO";
import { EmployeeServiceError } from "../error/employee.exception";

export class EmployeeService {
  private readonly takenCheffs = new Set<EmployeeDTO>();
  constructor(
    private readonly availableCheffs = new Set<EmployeeDTO>(),
    readonly availableWaiters = new Set<EmployeeDTO>()
  ) {}
  addCheffToCrew(name: string): void {
    const cheff = new EmployeeDTO(name, EmployeeRole.Cheff);
    this.availableCheffs.add(cheff);
  }
  getCheffToOrder(): EmployeeDTO {
    const cheff = Array.from(this.availableCheffs).reverse().pop();
    if (cheff) {
      this.takenCheffs.add(cheff);
      this.availableCheffs.delete(cheff);
      return cheff;
    }
    throw new EmployeeServiceError({
      name: "EMPLOYEE_ERROR",
      message: "There isn't free cheffs",
    });
  }
  isCheffAvailable(): boolean {
    return this.availableCheffs.size > 0;
  }
  relievedCheff(cheff: EmployeeDTO) {
    if (this.takenCheffs.delete(cheff)) {
      this.availableCheffs.add(cheff);
      return;
    }
    throw new EmployeeServiceError({
      name: "EMPLOYEE_ERROR",
      message: "This cheff wasn't taken",
    });
  }
}
