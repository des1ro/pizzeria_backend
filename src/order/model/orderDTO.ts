import { Discount } from "../../discounts/discount.enum";
import { PizzaDTO } from "../../pizza/model/pizzaDTO";

export class OrderDTO {
  constructor(
    readonly uuid: string,
    readonly discount: Discount,
    readonly products: PizzaDTO[],
    readonly tableId?: number
  ) {}
}
