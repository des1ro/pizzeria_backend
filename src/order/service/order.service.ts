import { EmployeeDTO } from "../../employee/model/employeeDTO";
import { PizzeriaError } from "../../exceptions/pizzeria.exceptions";
import { OrderDTO } from "../model/orderDTO";

export class OrderService {
  constructor(
    private readonly completedOrders = new Set<OrderDTO>(),
    private readonly orderQueque = new Set<OrderDTO>(),
    private readonly orderInProgress = new Map<OrderDTO, EmployeeDTO>()
  ) {}

  addOrderToQueque(order: OrderDTO): void {
    if (this.orderQueque.has(order)) {
      throw new PizzeriaError({
        name: "ORDER_ERROR",
        message: "Order is already in queque",
      });
    }
    this.orderQueque.add(order);
  }
  makeOrderInProgress(order: OrderDTO, cheff: EmployeeDTO) {
    if (this.orderQueque.delete(order)) {
      this.orderInProgress.set(order, cheff);
      return;
    }
    throw new PizzeriaError({
      name: "ORDER_ERROR",
      message: "There is no such as order in queque",
    });
  }
  completeOrderAndReturnCheff(order: OrderDTO): EmployeeDTO {
    const cheff = this.orderInProgress.get(order);
    if (this.orderInProgress.has(order) && cheff) {
      this.completedOrders.delete(order);
      this.completedOrders.add(order);
      return cheff;
    }
    throw new PizzeriaError({
      name: "ORDER_ERROR",
      message: "There is no order in order in progress",
    });
  }
  getOrderPrice(order: OrderDTO): number {
    return (
      (order.products.reduce((total, pizza) => total + pizza.price, 0) *
        (100 - order.discount)) /
      100
    );
  }
  getCompletedOrdersArray() {
    return Array.from(this.completedOrders);
  }
  getOrderId(): number {
    return (
      this.orderInProgress.size +
      this.orderQueque.size +
      this.completedOrders.size +
      1
    );
  }
}
