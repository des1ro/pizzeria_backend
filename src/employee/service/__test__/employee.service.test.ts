import { PizzeriaError } from "../../../exceptions/pizzeria.exceptions";
import { EmployeeRole } from "../../enum/employee.enum";
import { EmployeeDTO } from "../../model/employeeDTO";
import { EmployeeService } from "../employee.service";

describe("Employee service test suite", () => {
  let objectUnderTest: EmployeeService;
  let mockedEmployeeNameOne: string;
  let mockedEmployeeNameTwo: string;
  let mockedEmployeeNameThree: string;
  beforeAll(() => {
    mockedEmployeeNameOne = "test employee one";
    mockedEmployeeNameTwo = "test employee two";
    mockedEmployeeNameThree = "test employee three";
  });
  beforeEach(() => {
    objectUnderTest = new EmployeeService();
  });
  it("Should add cheff to crew", () => {
    //When
    objectUnderTest.addCheffToCrew(mockedEmployeeNameOne);
    //Then
    expect(objectUnderTest).toMatchSnapshot();
  });
  it("Should return which is firs on the list", () => {
    //Given
    objectUnderTest.addCheffToCrew(mockedEmployeeNameOne);
    objectUnderTest.addCheffToCrew(mockedEmployeeNameThree);
    objectUnderTest.addCheffToCrew(mockedEmployeeNameTwo);
    //When
    const resultOne = objectUnderTest.getCheffToOrder();
    const resultTwo = objectUnderTest.getCheffToOrder();
    const resultThree = objectUnderTest;
    //Then
    expect(resultOne).toMatchSnapshot();
    expect(resultTwo).toMatchSnapshot();
    expect(resultThree).toMatchSnapshot();
  });
  it("Should throw PizzeriaError if there isn't free cheff", () => {
    //Given
    objectUnderTest.addCheffToCrew(mockedEmployeeNameOne);
    //When
    const resultOne = objectUnderTest.getCheffToOrder();
    //Then
    expect(resultOne).toMatchSnapshot();
    expect(() => objectUnderTest.getCheffToOrder()).toThrow(PizzeriaError);
  });
  it("Should return true if cheff is available", () => {
    //Given
    objectUnderTest.addCheffToCrew(mockedEmployeeNameOne);
    //When
    const resultOne = objectUnderTest.isCheffAvailable();
    //Then
    expect(resultOne).toBe(true);
    expect(objectUnderTest).toMatchSnapshot();
  });
  it("Should return false if cheff isn't available", () => {
    //When
    const resultOne = objectUnderTest.isCheffAvailable();
    //Then
    expect(resultOne).toBe(false);
    expect(objectUnderTest).toMatchSnapshot();
  });
  it("Should relived cheff", () => {
    //Given
    objectUnderTest.addCheffToCrew(mockedEmployeeNameOne);
    //When
    const cheff = objectUnderTest.getCheffToOrder();
    objectUnderTest.relievedCheff(cheff);
    //Then
    expect(cheff).toMatchSnapshot();
    expect(objectUnderTest).toMatchSnapshot();
  });
  it("Should throw PizzeriaError if there is not cheff wasn't taken", () => {
    //Given
    objectUnderTest.addCheffToCrew(mockedEmployeeNameOne);
    const mockedCheff = new EmployeeDTO("test cheff", EmployeeRole.Cheff);
    //When
    const cheff = objectUnderTest.getCheffToOrder();
    objectUnderTest.relievedCheff(cheff);
    //Then
    expect(cheff).toMatchSnapshot();
    expect(objectUnderTest).toMatchSnapshot();
    expect(() => objectUnderTest.relievedCheff(mockedCheff)).toThrow(
      PizzeriaError
    );
  });
});
