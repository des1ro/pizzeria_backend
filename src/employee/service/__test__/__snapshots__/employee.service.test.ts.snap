// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Employee service test suite Should add cheff to crew 1`] = `
EmployeeService {
  "availableCheffs": Set {
    EmployeeDTO {
      "name": "test employee one",
      "role": "Cheff",
    },
  },
  "availableWaiters": Set {},
  "takenCheffs": Set {},
}
`;

exports[`Employee service test suite Should relived cheff 1`] = `
EmployeeDTO {
  "name": "test employee one",
  "role": "Cheff",
}
`;

exports[`Employee service test suite Should relived cheff 2`] = `
EmployeeService {
  "availableCheffs": Set {
    EmployeeDTO {
      "name": "test employee one",
      "role": "Cheff",
    },
  },
  "availableWaiters": Set {},
  "takenCheffs": Set {},
}
`;

exports[`Employee service test suite Should return false if cheff isn't available 1`] = `
EmployeeService {
  "availableCheffs": Set {},
  "availableWaiters": Set {},
  "takenCheffs": Set {},
}
`;

exports[`Employee service test suite Should return true if cheff is available 1`] = `
EmployeeService {
  "availableCheffs": Set {
    EmployeeDTO {
      "name": "test employee one",
      "role": "Cheff",
    },
  },
  "availableWaiters": Set {},
  "takenCheffs": Set {},
}
`;

exports[`Employee service test suite Should return which is firs on the list 1`] = `
EmployeeDTO {
  "name": "test employee one",
  "role": "Cheff",
}
`;

exports[`Employee service test suite Should return which is firs on the list 2`] = `
EmployeeDTO {
  "name": "test employee three",
  "role": "Cheff",
}
`;

exports[`Employee service test suite Should return which is firs on the list 3`] = `
EmployeeService {
  "availableCheffs": Set {
    EmployeeDTO {
      "name": "test employee two",
      "role": "Cheff",
    },
  },
  "availableWaiters": Set {},
  "takenCheffs": Set {
    EmployeeDTO {
      "name": "test employee one",
      "role": "Cheff",
    },
    EmployeeDTO {
      "name": "test employee three",
      "role": "Cheff",
    },
  },
}
`;

exports[`Employee service test suite Should throw EmployeeServiceError if there is not cheff wasn't taken 1`] = `
EmployeeDTO {
  "name": "test employee one",
  "role": "Cheff",
}
`;

exports[`Employee service test suite Should throw EmployeeServiceError if there is not cheff wasn't taken 2`] = `
EmployeeService {
  "availableCheffs": Set {
    EmployeeDTO {
      "name": "test employee one",
      "role": "Cheff",
    },
  },
  "availableWaiters": Set {},
  "takenCheffs": Set {},
}
`;

exports[`Employee service test suite Should throw EmployeeServiceError if there isn't free cheff 1`] = `
EmployeeDTO {
  "name": "test employee one",
  "role": "Cheff",
}
`;
